### PR TITLE
Improve dashboard charts - migrate to chart.js

### LINKED ISSUE #186: Improve dashboard charts
### Description (*)
Following https://github.com/magento/magento2/issues/21599, Magento recently switched from the deprecated Google Charts API to generating static chart images via image-charts.com (https://github.com/magento/magento2/commit/f92c8fd2b031531bac32615ab80ca09a66665617).

That being said, the new charts are not in any way dynamic, and do not provide a user experience befitting an enterprise software package, for reasons including unclear labels, lack of gridlines, pixelation, branding, and poor styling.

For example, this is the current dashboard chart on a site running 2.3.2 Commerce:
D-ZlZRuXsAMrRqp

### Expected behavior (*)
Use a modern JavaScript chart library to build the dashboard charts on the client side.

### Benefits
Using a client-side JS chart library would provide a far superior user experience than the current static images, in design, interactivity (tooltips), and usefulness.

Doing so would also reducing any possible security or privacy concerns from the current approach, which involves sending data to or including a third-party-hosted resource or API for the same purpose. With a client-side library responsible for rendering the data, no data would leave the premises, and no external resources would be added in to the sensitive admin panel area.

### Additional information
There are numerous open-source JS chart libraries that would fit the purpose, including https://github.com/chartjs/Chart.js and https://github.com/flot/flot (both licensed MIT). My inclination would be Chart.js, given its more active contributor base, better documentation, and simple out-of-box responsive support.

In the process, the charts should also be switched from line to bar-type, to better reflect the nature of the underlying data (which is discrete sales figures, not continuous).

After integrating such a library, further improvements would be possible in the areas of changing timescale without reload; breaking down data by store; and more.

### PR DESCRIPTION
**Pull request by Bartlomiejsz**
### Description (*)

This PR aims to migrate dashboard charts to chart.js library instead of static images.
Fixes magento/community-features#186
This is preview of new charts:
image

### Manual testing scenarios (*)

1. Enable Dashboard Charts
2. Place some orders with different dates and amounts, process them to `Complete` status
3. Go to MBO Dashboard
4. Dynamic charts should be visible 

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, here is your [new Magento instance](https://pr-26923.instances.magento-community.engineering).
Admin access: [https://pr-26923.instances.magento-community.engineering/admin_6356](https://pr-26923.instances.magento-community.engineering/admin_6356)
Login: `63a15b5d` Password: `a49480ba054f`
Instance will be terminated in up to 3 hours.

**Comment by ihor-sviziev** [comment]
@Bartlomiejsz looks really promising, however you've reused/removed big parts of code that was used originally, some of these parts marked as "@api", so they can't contain backward incompatible changes.
In places where the most logic was changed probably it's better to create new class and deprecate older ones.

File: app/code/Magento/Backend/Controller/Adminhtml/Dashboard/Graph/Amounts.php
Comment (ihor-sviziev) [inline app/code/Magento/Backend/Controller/Adminhtml/Dashboard/Graph/Amounts.php]: Could you add specific factory to constructor and use it there?
There were long discussion for this some time ago, you can see more details here:
https://github.com/magento/magento2/pull/24711#issuecomment-542469839

**Comment by Bartlomiejsz** [comment]
Sure @ihor-sviziev, I'll work on moving new functionalities to newly create classes, thank you for review.
Got question, it seems that Static Tests are using different rules for CE edition, and different for B2B - there are only two static issues for CE, but multiple for B2B - should I then fix all of those issues, or that means that B2B module configuration is not updated to latest version and I can ignore that?

**Comment by ihor-sviziev** [comment]
@Bartlomiejsz 
Here is answer for your question :) 
https://github.com/magento/magento2/pull/26923/checks?check_run_id=454975377
image

File: app/code/Magento/Backend/Controller/Adminhtml/Dashboard/Graph/Amounts.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Controller/Adminhtml/Dashboard/Graph/Amounts.php]: Done @ihor-sviziev

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Backend/Block/Dashboard/Chart.php
Comment (ihor-sviziev) [inline app/code/Magento/Backend/Block/Dashboard/Chart.php]:
```suggestion
    private $htmlId = '';
```

File: app/code/Magento/Backend/Block/Dashboard/Chart.php
Comment (ihor-sviziev) [inline app/code/Magento/Backend/Block/Dashboard/Chart.php]:
```suggestion
    public function setHtmlId(string $htmlId): void
```

File: app/code/Magento/Backend/Block/Dashboard/Chart.php
Comment (ihor-sviziev) [inline app/code/Magento/Backend/Block/Dashboard/Chart.php]:
```suggestion
    public function setUpdateUrl(string $updateUrl): void
```

**Comment by Bartlomiejsz** [comment]
@ihor-sviziev I believe it's done, please review :)

File: app/code/Magento/Backend/Block/Dashboard/Grids.php
Comment (ihor-sviziev) [inline app/code/Magento/Backend/Block/Dashboard/Grids.php:16]: Why do you think it should be marked as API?

File: app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php
Comment (ihor-sviziev) [inline app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php:16]: Why do you think it should be marked as API ?

File: app/code/Magento/Backend/Controller/Adminhtml/Dashboard/AjaxBlock.php
Comment (ihor-sviziev) [inline app/code/Magento/Backend/Controller/Adminhtml/Dashboard/AjaxBlock.php:59]: Look like after this change we don’t need anymore dynamic class name generation

File: app/code/Magento/Backend/Controller/Adminhtml/Dashboard/AjaxBlock.php
Comment (ihor-sviziev) [inline app/code/Magento/Backend/Controller/Adminhtml/Dashboard/AjaxBlock.php]: Please revert this change
We already had discussion for that, decided to use specific factories
https://github.com/magento/magento2/pull/24711#issuecomment-535107638

File: app/code/Magento/Backend/view/adminhtml/templates/dashboard/chart.phtml
Comment (ihor-sviziev) [inline app/code/Magento/Backend/view/adminhtml/templates/dashboard/chart.phtml]: Looks like we need to escape output here

File: app/code/Magento/Backend/view/adminhtml/templates/dashboard/totalbar.phtml
Comment (ihor-sviziev) [inline app/code/Magento/Backend/view/adminhtml/templates/dashboard/totalbar.phtml]:
```suggestion
            <?php foreach ($block->getTotals() as $total): ?>
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Backend/Block/Dashboard/Grids.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Block/Dashboard/Grids.php:16]: @ihor-sviziev after I moved adding dashboard blocks fro `_prepareLayout` method to xml files, I got failure in (i think) static tests that I added block that is not marked as api to layout. So I was sure this is requirement in that case. What should I do with this?

File: app/code/Magento/Backend/Block/Dashboard/Chart.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Block/Dashboard/Chart.php]: Resolving since this file no longer exists

File: app/code/Magento/Backend/Block/Dashboard/Chart.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Block/Dashboard/Chart.php]: Resolving since this file no longer exists

File: app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php:16]: @ihor-sviziev after I moved adding dashboard blocks fro `_prepareLayout` method to xml files, I got failure in (i think) static tests that I added block that is not marked as api to layout. So I was sure this is requirement in that case. What should I do with this?

File: app/code/Magento/Backend/Controller/Adminhtml/Dashboard/AjaxBlock.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Controller/Adminhtml/Dashboard/AjaxBlock.php:59]: @ihor-sviziev done

File: app/code/Magento/Backend/Controller/Adminhtml/Dashboard/AjaxBlock.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Controller/Adminhtml/Dashboard/AjaxBlock.php]: @ihor-sviziev done

File: app/code/Magento/Backend/view/adminhtml/templates/dashboard/chart.phtml
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/view/adminhtml/templates/dashboard/chart.phtml]: @ihor-sviziev done

File: app/code/Magento/Backend/view/adminhtml/templates/dashboard/totalbar.phtml
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/view/adminhtml/templates/dashboard/totalbar.phtml]: @ihor-sviziev done

File: app/code/Magento/Backend/Block/Dashboard/Chart.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Block/Dashboard/Chart.php]: Resolving since this file no longer exists

File: app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php
Comment (ihor-sviziev) [reply app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php:16]: Could you point me to test result? It doesn't looks to me as expected behavior

File: app/code/Magento/Backend/view/adminhtml/templates/dashboard/totalbar.phtml
Comment (ihor-sviziev) [inline app/code/Magento/Backend/view/adminhtml/templates/dashboard/totalbar.phtml:12]: Hi @Bartlomiejsz,
I'm not sure I've ever saw such example. Could you add more details how it works, or where can I read more details?

Comment (Bartlomiejsz) [reply app/code/Magento/Backend/view/adminhtml/templates/dashboard/totalbar.phtml:12]: Hi @ihor-sviziev, escaper methods in AbstractBlock were deprecated some time ago, please check this commit: https://github.com/magento/magento2/commit/b5fde42dacbca4239e9276f871fc6b48840a28a7#diff-0e10c44d6b0961dcab05891b110d0305

File: app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php:16]: I faced this when this PR was draft, I force pushed changes later. I can remove one annotation and push the changes for testing purpose

File: app/code/Magento/Backend/view/adminhtml/templates/dashboard/totalbar.phtml
Comment (ihor-sviziev) [reply app/code/Magento/Backend/view/adminhtml/templates/dashboard/totalbar.phtml:12]: Thank you for pointing me! Looks really good!

File: app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php:16]: @ihor-sviziev here it comes https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/26923/3019ba1dfb621066ed1bd93a177981b8/Statics/allure-report-b2b/index.html#categories/9f143fbe76131fd0113101cfa4fa3008/85c8b92bfa927797/

Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php:16]: image
Posting image also since this report might vanish later when I'll force push reverting last commit

**Comment by ihor-sviziev** [comment]
@lenaorobei 
Could you help with this test failure? 
As for me it should be ignored. I actually don't really know how to do that.
https://github.com/magento/magento2/pull/26923#discussion_r383949720

**Comment by ihor-sviziev** [comment]
As we discussed with @lenaorobei - we need to mark blocks with `@api` tag.
Here is description of static test:
```
     * Since blocks can be referenced from templates, they should be stable not to break theme customizations.
     * So all blocks should be @api annotated. This test checks that all blocks declared in layout files are public
```

File: app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php
Comment (ihor-sviziev) [reply app/code/Magento/Backend/Block/Dashboard/Orders/Grid.php:16]: @Bartlomiejsz we discussed it, blocks should be marked as `@api`, you did everything correct last time. That was something new for me, sorry for confusing

**Comment by Bartlomiejsz** [comment]
Done @ihor-sviziev

File: app/code/Magento/Backend/Block/Dashboard/Grids.php
Comment (ihor-sviziev) [reply app/code/Magento/Backend/Block/Dashboard/Grids.php:16]: As we discussed - it's ok.

**Comment by ihor-sviziev** [comment]
@magento run Static Tests, Semantic Version Checker

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6988](https://jira.corp.magento.com/browse/ENGCOM-6988) has been created to process this Pull Request

**Comment by lenaorobei** [comment]
Talked with PO @misha-kotov and got green light for this PR 🎉 
@Bartlomiejsz that's an awesome improvement!

**Comment by Bartlomiejsz** [comment]
Thank you @lenaorobei :slightly_smiling_face:

File: app/code/Magento/Backend/Model/Dashboard/Chart/Date.php
Comment (VladimirZaets) [inline app/code/Magento/Backend/Model/Dashboard/Chart/Date.php:77]: Why you don't use TimezoneInterface in this case. It can convert/format data according to timezone?

File: app/code/Magento/Backend/Model/Dashboard/Chart.php
Comment (VladimirZaets) [inline app/code/Magento/Backend/Model/Dashboard/Chart.php]: Use request in the model is bad practice. Request objects should be used just in the controller. The model is another layer and it shouldn't be depended to the HTTP request

File: lib/web/chartjs/Chart.js
Comment (VladimirZaets) [inline lib/web/chartjs/Chart.js]: I think we don't need full version of the library. I think the bundled/minified version is enough for us. According to our rules, we can't change 3rd party libraries ( because we will haven't the ability to update them), we use 3rd party libraries as "black box", and operate just with the library by contract. In this case, we don't need full versions

**Review by VladimirZaets — CHANGES_REQUESTED** [review]
Hi @Bartlomiejsz. Thanks for collaboration. Please look at my review comments.

File: app/code/Magento/Backend/Model/Dashboard/Chart/Date.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Model/Dashboard/Chart/Date.php]: @VladimirZaets actually this part is not something I developed, this was just moved from current graph implementation into its own model. I could try to adjust it to use TimezoneInterface if this is required though

Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Model/Dashboard/Chart/Date.php]: Hmm, actually it is using TimezoneInterface, what change exactly you mean here?

**Comment by naydav** [comment]
@Bartlomiejsz  @lenaorobei 

Hello guys,
Please evaluate the option of extracting new functionality in a separate module.
Also, it will be great if we will introduce some Interfaces for extensibility and customization of current behavior.

Thanks

**Review by naydav — APPROVED** [review]

**Review by naydav — CHANGES_REQUESTED** [review]
See my comment above

**Comment by magento-engcom-team** [comment]
Hi @naydav, thank you for the review.
[ENGCOM-6988](https://jira.corp.magento.com/browse/ENGCOM-6988) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@naydav not sure it really require creating separate module and don’t thing anyone would want to replace implementation to own.
This feature request was to make charts just better, right now they’re ugly.
Let’s finalize and merge this PR till we could include it into 2.4.0, otherwise such cool feature will not be available

File: app/code/Magento/Backend/Model/Dashboard/Chart.php
Comment (ihor-sviziev) [reply app/code/Magento/Backend/Model/Dashboard/Chart.php]: How would you like to pass it to the model?

File: lib/web/chartjs/Chart.js
Comment (ihor-sviziev) [reply lib/web/chartjs/Chart.js]: Good point, agree with it

**Review by melnikovi — DISMISSED** [review]

**Comment by naydav** [comment]
Hi @ihor-sviziev,
If you see that there is no way to do this easily, I have no objections to merging it.
Thanks

**Review by naydav — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @naydav, thank you for the review.
[ENGCOM-6988](https://jira.corp.magento.com/browse/ENGCOM-6988) has been created to process this Pull Request

File: app/code/Magento/Backend/Model/Dashboard/Chart.php
Comment (VladimirZaets) [reply app/code/Magento/Backend/Model/Dashboard/Chart.php]: @ihor-sviziev controller should pass all required data to the method

**Comment by ihor-sviziev** [comment]
@naydav extracting some functionality to external module by definition not easy thing. I don’t believe there any need for that, as I didn’t saw any reported issues related to missing extension points in these parts

File: app/code/Magento/Backend/Model/Dashboard/Chart/Date.php
Comment (edenduong) [inline app/code/Magento/Backend/Model/Dashboard/Chart/Date.php]: I think we should get the value from `const`

File: app/code/Magento/Backend/ViewModel/ChartDisabled.php
Comment (edenduong) [inline app/code/Magento/Backend/ViewModel/ChartDisabled.php]: It should be changed to get value from `const`

**Review by edenduong — CHANGES_REQUESTED** [review]
I think we should change the value like '7d' to get value from`const` or some config xml path should be `const` also.

File: lib/web/chartjs/Chart.js
Comment (Bartlomiejsz) [reply lib/web/chartjs/Chart.js]: Done, I removed all files but minified js required to use charts

File: app/code/Magento/Backend/Model/Dashboard/Chart.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Model/Dashboard/Chart.php]: Done

File: app/code/Magento/Backend/Model/Dashboard/Chart/Date.php
Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Model/Dashboard/Chart/Date.php:77]: Hi @VladimirZaets, could you please explain that? I didn't developed this part, it was just moved from old Graph block implementation to it's own model. Also, those dates objects are coming from `Magento\Reports\Model\ResourceModel\Order\Collection` and have timezone set

File: app/code/Magento/Backend/Model/Dashboard/Chart/Date.php
Comment (ihor-sviziev) [reply app/code/Magento/Backend/Model/Dashboard/Chart/Date.php]: @Bartlomiejsz could you extract all ranges to constants?

File: app/code/Magento/Backend/Model/Dashboard/Chart.php
Comment (ihor-sviziev) [inline app/code/Magento/Backend/Model/Dashboard/Chart.php]: I think it’s good idea to define argument types in all new classes. Will you be able to update your changes?

Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Model/Dashboard/Chart.php]: sure, I'll update those. What about changes mentioned by @edenduong, should I move some parts to constants? Keep in mind that i.e. periods are used in few other places I didn't modified yet, and it's hardcoded in all of those, so this won't be easily extendable without rebuilding few other classes.

Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Model/Dashboard/Chart.php]: Also functional tests are failing but this is happening since few days to all of my PRs even if I didn't modified anything related - seems those are flaky

Comment (ihor-sviziev) [reply app/code/Magento/Backend/Model/Dashboard/Chart.php]: Can’t we just create all period constants in one class / interface and use in all places ?

Comment (Bartlomiejsz) [reply app/code/Magento/Backend/Model/Dashboard/Chart.php]: Sure, that's of course possible. I'll just have to verify few others not yet touched places, I will try to do it today later

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Integration Tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6988](https://jira.corp.magento.com/browse/ENGCOM-6988) has been created to process this Pull Request

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-6988](https://jira.corp.magento.com/browse/ENGCOM-6988) has been created to process this Pull Request

**Review by edenduong — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @edenduong, thank you for the review.
[ENGCOM-6988](https://jira.corp.magento.com/browse/ENGCOM-6988) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Comment by lenaorobei** [comment]
Waiting for internal approval request [MC-31973](https://jira.corp.magento.com/browse/MC-31973).

**Comment by lenaorobei** [comment]
Created a related DevDocs ticket https://github.com/magento/devdocs/issues/6830

**Comment by lenaorobei** [comment]
Hi @Bartlomiejsz! Can I ask you to improve `AdminDashboardWithChartsTest` that covers this functionality? We need to:
- fix description
- change `testCaseId` and `useCaseId` to `MC-28918`
- add few more checks
   - Test Step: Open Dashboard, assert charts
   - Orders and Amounts tab; Drop-down: Last 24 Hours, Last 7 Days, Current Month, YTD, 2 YTD
   - Expected result: Quantity increased for 1 order, Amounts - according to payment: *subtotal (price  * qty) + tax + shipping - discounts (or gift cards if applied) = base_grand_total*
(Orders tab shows data of orders count, Amounts shows data of Revenue value)

**Comment by Bartlomiejsz** [comment]
Hi @lenaorobei,
I can work on that, but I'm not sure if I understand last point correctly (Orders tab shows data of orders count, Amounts shows data of Revenue value) - this can be checked on chart, but chart is rendered on canvas by chart.js, so I don't think MFTF test have possibility to verify what is rendered there.
Could you please specify what exactly should be checked here?

**Comment by lenaorobei** [comment]
@Bartlomiejsz, you are right, I just double checked that. 
Please just ignore my comment 🙂 The test that was supposed to test charts is not related to the charts at all.

**Comment by Bartlomiejsz** [comment]
@lenaorobei ok, I can check if there is possibility to add some valuable assertions there, but otherwise I will just not touch it :)

**Comment by Bartlomiejsz** [comment]
@lenaorobei from what I can see now, it seems it is related, at least was :) It was checking if after placing an order image chart is displayed. So this may indeed require some adjustments after my changes, I'll try to do that, will keep you posted in case of issues

**Comment by lenaorobei** [comment]
[MC-31973](https://jira.corp.magento.com/browse/MC-31973) is approved and we are ready to move forward with it.

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-6988](https://jira.corp.magento.com/browse/ENGCOM-6988) has been created to process this Pull Request

**Comment by Bartlomiejsz** [comment]
Hi @lenaorobei, in that case changes to this single MFTF test I should probably introduce as separate PR once this will be merged, am I right?

**Comment by lenaorobei** [comment]
@Bartlomiejsz yes, let's do separate PR. Thanks.

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by lenaorobei** [comment]
@mageno run Integration Tests

**Comment by lenaorobei** [comment]
@magento, run Integration Tests

**Comment by lenaorobei** [comment]
@magento run Integration Tests

**Comment by ihor-sviziev** [comment]
Yeah! It was finally merged! That’s amazing contribution @Bartlomiejsz! Thank you so much!

**Comment by Bartlomiejsz** [comment]
Thanks @ihor-sviziev and thank you for your help on this!

**Pull request merged**
